snapshots/btrfs/btrfs.go

```diff
@@ -127,19 +127,57 @@ func (b *snapshotter) Update(ctx context.Context, info snapshots.Info, fieldpath
 }
 
 // Usage retrieves the disk usage of the top-level snapshot.
+//
+// For active snapshot, the `excl` value of L0 QGroup is used, and committed on Commit.
+// For view snapshot, always 0 is returned.
+// For committed snapshot, the committed value is returned.
 func (b *snapshotter) Usage(ctx context.Context, key string) (snapshots.Usage, error) {
-	panic("not implemented")
-
-	// TODO(stevvooe): Btrfs has a quota model where data can be exclusive to a
-	// snapshot or shared among other resources. We may find that this is the
-	// correct value to reoprt but the stability of the implementation is under
-	// question.
-	//
-	// In general, this has impact on the model we choose for reporting usage.
-	// Ideally, the value should allow aggregration. For overlay, this is
-	// simple since we can scan the diff directory to get a unique value. This
-	// breaks down when start looking the behavior when data is shared between
-	// snapshots, such as that for btrfs.
+	ctx, t, err := b.ms.TransactionContext(ctx, false)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	defer t.Rollback()
+	id, info, committedUsage, err := storage.GetInfo(ctx, key)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	switch info.Kind {
+	case snapshots.KindActive:
+		return b.activeUsage(ctx, id)
+	case snapshots.KindView:
+		return snapshots.Usage{Size: 0, Inodes: 0}, nil
+	default:
+		return committedUsage, nil
+	}
+
+}
+
+// activeUsage returns the `excl` value of L0 QGroup is used, and committed on Commit.
+func (b *snapshotter) activeUsage(ctx context.Context, activeSnapshotID string) (snapshots.Usage, error) {
+	target := filepath.Join(b.root, "active", activeSnapshotID)
+	subvolID, err := btrfs.SubvolID(target)
+	if err != nil {
+		return snapshots.Usage{}, err
+	}
+	if err := btrfs.Sync(target); err != nil {
```

Comment:
How confident are we that after this Sync is called the info is accurate?

Reply:
As accurate as `btrfs qgroup show --sync`:  https://github.com/kdave/btrfs-progs/blob/9fe889ac02b9c49b885c8999f5dd4e192697fa83/cmds-qgroup.c#L377-L382

man: http://manpages.ubuntu.com/manpages/zesty/man8/btrfs-qgroup.8.html
```
    --sync
               To retrieve information after updating the state of qgroups,
               force sync of the filesystem identified by <path> before
               getting information.
```